### PR TITLE
Scripting: Add watcher script contexts

### PR DESCRIPTION
**Pull request by rjernst**
This commit removes the use of ExecutableScript from watcher in favor of
custom script contexts for both watcher condition scripts and transform
scripts.

### CODE REVIEW

File: x-pack/plugin/watcher/src/main/java/org/elasticsearch/xpack/watcher/support/Variables.java [hunk 1]
```diff
@@ -22,7 +22,13 @@
...
     public static final String VARS = "vars";
 
-    public static Map<String, Object> createCtxModel(WatchExecutionContext ctx, Payload payload) {
+    public static Map<String, Object> createCtxParamsMap(WatchExecutionContext ctx, Payload payload) {
+        Map<String, Object> model = new HashMap<>();
+        model.put(CTX, createCtx(ctx, payload));
+        return model;
+    }
+
+    public static Map<String, Object> createCtx(WatchExecutionContext ctx, Payload payload) {
```
Comment (hub-cap) [inline x-pack/plugin/watcher/src/main/java/org/elasticsearch/xpack/watcher/support/Variables.java]: do we still need this? should it be @deprecated?

**Review by hub-cap — COMMENTED** [review]
These changes look fine to me, but i wouldnt mind @spinscale 	also taking a look. I dont know much about Scripts in watcher.

File: x-pack/plugin/watcher/src/main/java/org/elasticsearch/xpack/watcher/support/Variables.java (hunk 1 above)
Comment (rjernst) [reply x-pack/plugin/watcher/src/main/java/org/elasticsearch/xpack/watcher/support/Variables.java]: Not sure what you mean. This was extracted from the previous method to allow only getting the actual ctx map, but not putting it into an outer params map (the new contexts need this so they can put it into the wrapped ParameterMap for deprecation warnings).

Comment (hub-cap) [reply x-pack/plugin/watcher/src/main/java/org/elasticsearch/xpack/watcher/support/Variables.java]: yea i just mean if we dont want coders (me and @spinscale) accidentally using the createCtx method, u could private it, or if other things are still using it, mark it @deprecated in favor of using the other method `createCtxParamsMap`

Comment (rjernst) [reply x-pack/plugin/watcher/src/main/java/org/elasticsearch/xpack/watcher/support/Variables.java]: Both are used for now. I will add a comment to try to make it clearer which should be used when.

Comment (spinscale) [reply x-pack/plugin/watcher/src/main/java/org/elasticsearch/xpack/watcher/support/Variables.java]: could you have just a second `createCtxParamsMap` method that takes another map as parameter? Which is an empty hash map by default used to fill and the parmaters in `WatcherConditionScript`?

Comment (rjernst) [reply x-pack/plugin/watcher/src/main/java/org/elasticsearch/xpack/watcher/support/Variables.java]: I think having 2 methods with the same name would be more confusing. Do the java docs I added not make sense?

File: x-pack/plugin/watcher/src/main/java/org/elasticsearch/xpack/watcher/transform/script/WatcherTransformScript.java [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+public abstract class WatcherTransformScript {
+    public static final String[] PARAMETERS = {};
+
+    private static final Map<String, String> DEPRECATIONS;
+
+    static {
+        Map<String, String> deprecations = new HashMap<>();
+        deprecations.put(
+            "ctx",
+            "Accessing variable [ctx] via [params.ctx] from within a watcher_transform script " +
```
Comment (spinscale) [inline x-pack/plugin/watcher/src/main/java/org/elasticsearch/xpack/watcher/transform/script/WatcherTransformScript.java]: is it worth testing this somewhere or did I miss the test? Same applies for the `WatcherConditionScript`

**Review by spinscale — APPROVED** [review]
My Script Context skills are not the best, from a watcher perspective this looks good to me though.

Left one question regarding adding tests for the deprecation.

File: x-pack/plugin/watcher/src/main/java/org/elasticsearch/xpack/watcher/condition/WatcherConditionScript.java [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * A script to determine whether a watch should be run.
+ */
+public abstract class WatcherConditionScript {
+    public static final String[] PARAMETERS = {};
```
[the remark this answers is not in the document]
Comment (rjernst) [reply x-pack/plugin/watcher/src/main/java/org/elasticsearch/xpack/watcher/condition/WatcherConditionScript.java]: This constant must exist for all script classes, even if the execute method takes zero arguments.

File: x-pack/plugin/watcher/src/main/java/org/elasticsearch/xpack/watcher/transform/script/WatcherTransformScript.java (hunk 2 above)
Comment (rjernst) [reply x-pack/plugin/watcher/src/main/java/org/elasticsearch/xpack/watcher/transform/script/WatcherTransformScript.java]: I added tests in e1e0490

**Review by spinscale — APPROVED** [review]

**Comment by ypid-geberit** [comment]
For reference, this PR prevented me to use the same Painless script for both the watch condition and transformation because the condition now enforces the return type to be a boolean.

Ref: https://discuss.elastic.co/t/regression-es-6-2-6-5-watch-condition-cannot-cast-from-java-util-hashmap-to-boolean/161180

The Elastic support was kind enough to work out a solution with me which still allows to only use one script for both:

```YAML
---

# yamllint disable rule:line-length rule:comments-indentation

trigger:
  schedule:
    hourly:
      minute: 32

input:
  search:
    timeout: '5m'
    request:

      indices:
        - '*-*'
        - '-.*'

      body:

        size: 0
        query:
          bool:
            must:
              - exists:
                  field: '@timestamp'

condition:
  script:
    inline: |
      ctx.payload.foo = 'testing';
      return false;

# transform:
#   script:
#     inline: |
#       ctx.payload.foo = 'testing';
#       return true;

actions:
  log:
    logging:
      level: info
      text: 'Watch payload after transform: {{ctx.payload}}'
```

**Pull request merged**